tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala

```diff
@@ -264,8 +267,25 @@ final class ScalaJSClassEmitter private (
       val js.Function(params, body) = genConstructorFunForJSClass(tree)
       js.MethodDef(static = false, js.Ident("constructor"), params, body)
     } else {
+      val className = tree.name.name
       val fieldDefs = genFieldDefs(tree)
-      if (fieldDefs.isEmpty && outputMode == OutputMode.ECMAScript6) {
+
+      val (otherStatements, args) = if (usesJSConstructorOpt(className)) {
+        val (head :: _) =
+          tree.memberMethods.map(_.tree).
+          filter(x => isConstructorName(x.name.name))
+        val js.Function(args, bodyBlock) = desugarToConstructor(this,
+            tree.name.name, head.args, head.body, js.Skip())
+        val body = bodyBlock match {
+          case js.Block(bdy) => bdy
+          case _             => Nil
+        }
+        (body, args)
+      } else {
+        (fieldDefs, Nil)
+      }
```

Comment:
There's something wrong. You shouldn't throw away `fieldDefs` in the case `usesJSConstructorOpt`. The `fieldDefs` should always be present. They ensure fields are all created and initialized with the zero of their type before starting the body of the constructor.

Consider returning `(js.Skip(), Nil)` in the else branch. And conclude with `js.Block(superCtorCall :: initClassData :: fieldDefs ::: otherStatements :: Nil)`.